src/combinatorics/composition.lean

```diff
@@ -0,0 +1,584 @@
+/-
+Ccompoyright (c) 2020 Sébastien Gouëzel. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Sébastien Gouëzel
+-/
+
+import data.fintype.card tactic.omega tactic.tidy
+
+/-!
+# Compositions
+
+A composition of an integer `n` is a decomposition of `{0, ..., n-1}` into blocks of consecutive
+integers. Equivalently, it is a decomposition `n = i₀ + ... + i_{k-1}` into a sum of positive
+integers, where the `iⱼ` are the lengths of the blocks. This notion is closely related to that
+of a partition of `n`, but in a composition of `n` the order of the `iⱼ`s matters.
+
+We represent a composition of `n` as a subset of `{0, ..., n}` containing `0` and `n`, where
+the elements of the subset (but `n`) correspond to the leftmost points of each block. Another
+implementation would be a subset of `{0, ..., n-1}` containing `0`, but this fails in the edge case
+`n = 0`. Another implementation would be as an integer `k` and a function `fin k → ℕ` summing to
+`n`, but this turns out to be less handy because of dependent type issues.
+
```

Comment:
Couldn't we just do arbitrary subsets of `fin (n-1)`, where the elements of the subset correspond to rightmost points of non-terminal blocks?

* for n=0, `fin (0-1) = fin 0 = empty`, and there's a unique subset corresponding to the unique composition `0 = "the empty sum"`
* for n=1, `fin (1-1) = fin 0 = empty`, and there's a unique subset corresponding to the unique composition `1 = 1`
* for n=2, `fin (2-1) = fin 1 = {0}`, and there are two subsets `{}` corresponding to `2 = 2` and `{0}` corresponding to the block decomposition of `[0,1]` as `[[0],[1]]`, in which the only non-terminal block is `[0]`, and its rightmost point is `0` as specified.
* etc

I haven't actually read what you do below, so this may actually be a terrible implementation. But at least the cardinality statement should be trivial. :-)

Reply:
Okay, I now see that `composition_equiv` below is exactly this. I'm certainly not suggesting you change anything, but perhaps this paragraph could give a better account of why you want this representation, without mentioning incorrect ones.

Reply:
Done.

Reply:
I'm struggling try to review this one, as it feels like everything is just dealing the pain of transforming between different representations, and I don't know what is inevitable pain and what is pain worth avoiding. Perhaps you could say more, either here or on Zulip, why you avoided what seems the "obvious" definition:

```lean
def composition (n : ℕ) : Type :=
{ blocks : list ℕ+ // (blocks.map (λ n : ℕ+, (n : ℕ))).sum = n}
```

Reply:
I spent a while trying to use this definition. I didn't get very far (trying to simultaneously homeschool kids :-) but it was [looking okay](https://gist.github.com/semorrison/191c7bc2d6847c26c7f311aeaea94e04) to begin with. (That gist depends on another file with lots of "missing API" about `list`, so it's not directly compilable.)

Reply:
The main player I want to use in the end is the embedding of `fin (c.size i)` in `fin n` corresponding to the `i`th block, because that's what I use to define the composition of analytic functions in #2328. This is why I used the definition in terms of subset of `fin n.succ`. It would also work perfectly fine with your definition, and in fact I started with a definition like that. I switched because I wasn't comfortable with using lists, while I know how to use finsets.

Reply:
In fact I tried with a structure containing `k` and a map `fin k -> pnat`. This was bad for DTT hell reasons. The list idea might solve this, so I could give it a shot. Could you share your file adding a good list API?

Reply:
Also, I need the fact that the cardinality is `2^(n-1)`. If I start with your definition, I will need to show that there's an equiv with something akin to my definition, and to construct the equiv I will need to do what I do in the PR. I don't see an obvious shortcut here.

Reply:
I will try to move it to lists, it would be much more convenient for some applications. Marking this as WIP